PhysicsPackages/fixedSourcePhysicsPackage_class.f90

```diff
@@ -189,11 +192,31 @@ subroutine cycles(self, tally, N_cycles)
             if(p % isDead) exit history
           end do history
 
+          ! If buffer is quite full, shift some particles to the commonBuffer
+          if (associated(self % commonBuffer) .and. (buffer % popSize() > self % bufferShift)) then
+            bufferExtra = buffer % popSize() - self % bufferShift
+            do j = 1, bufferExtra
+              call buffer % release(transferP)
+              call self % commonBuffer % detainCritical(transferP)
+            end do
+          end if
+
           ! Clear out buffer
-          if (buffer % isEmpty()) then
-            exit bufferLoop
-          else
+          if (.not. buffer % isEmpty()) then
             call buffer % release(p)
+            cycle
+          end if
+
+          ! Clear out common queue
+          if (associated(self % commonBuffer)) then
+            !$omp critical
+            if (.not. self % commonBuffer % isEmpty()) then
+              call self % commonBuffer % releaseCritical(p)
+            end if
+            !$omp end critical
```

Comment:
```suggestion
            if (.not. self % commonBuffer % isEmpty()) then
              call self % commonBuffer % releaseCritical(p)
            end if
```
The critical region is controlled inside the subroutine isn't it?

Reply:
The thing I'm being careful about here is that check on whether the buffer is empty. If there are two threads and one particle in the buffer, both will enter the conditional, the first will critically remove the particle, while the second will then be left trying to pop a particle from an empty buffer.
We could put a conditional inside the release as a check in principle? But we would still have the critical in the physics package too. What do you think?

Reply:
Good point! It completely eluded me. I guess the only thing that is worth changing here is maybe adding a comment to explain why it is like that. I am sure I am not a last person who might be confused by it :-)!

Reply:
Hear hear... We only found out because we were bitten by it ourselves!